File: su/src/main/java/org/sqlbuilder/su/Kb.java
package org.sqlbuilder.su;

import com.articulate.sigma.Formula;
import com.articulate.sigma.KB;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

public class Kb extends KB implements Serializable
{
	private static final long serialVersionUID = 3120000480284537868L;

	private static final String[] CORE_FILES = new String[]{"Merge.kif", "Mid-level-ontology.kif", "english_format.kif"};

	private static final boolean silent = System.getProperties().containsKey("SILENT");

	private String[] filenames;

	public Kb(final String dirName)
	{
		super("SUMO", dirName);
	}

	public boolean make(final boolean full)
	{
		make(Kb.getFiles(this.kbDir, full));
		return true;
	}

	public boolean make(final String[] files)
	{
		this.filenames = files != null ? files : Kb.getFiles(this.kbDir, true);
		final String[] filePaths = new String[this.filenames.length];
		for (int i = 0; i < filePaths.length; i++)
		{
			filePaths[i] = this.kbDir + File.separatorChar + this.filenames[i];
		}
		Kb.makeKB(this, filePaths);
		return true;
	}

	private static void makeKB(final KB kb, final String[] filePaths)
	{
		for (final String filePath : filePaths)
		{
			System.out.println("\n" + filePath);
			kb.addConstituent(filePath);
		}
	}

	public boolean makeClausalForms()
	{
		long count = 0;
		for (Collection<Formula> fs : formulaIndex.values())
		{
			for (Formula f : fs)
			{
				/* Tuple.Triple<List<Clause>, Formula, Map<String, String>> cf = */
				f.getClausalForms();
				if (!silent)
				{
					if ((count++ % 100L) == 0)
					{
						System.out.println();
					}
					System.out.print('.');
				}
			}
		}
		return true;
	}

	protected static String[] getFiles(final String dirName, final boolean full)
	{
		if (full)
		{
			final List<String> list = new ArrayList<>(Arrays.asList(Kb.CORE_FILES));
			for (final String filename : Kb.getKifs(dirName))
			{
				if (list.contains(filename))
				{
					continue;
				}
				list.add(filename);
			}
			return list.toArray(new String[0]);
		}
		return Kb.CORE_FILES;
	}

	private static String[] getKifs(final String dirName)
	{
		final File file = new File(dirName);
		if (file.exists() && file.isDirectory())
		{
			return file.list((dir, name) -> name.endsWith(".kif"));
		}
		return new String[]{};
	}

	public String[] getFilenames()
	{
		return this.filenames;
	}
}

File: su/src/main/java/org/sqlbuilder/su/Kb.kt
package org.sqlbuilder.su

import org.sigma.core.KB
import org.sqlbuilder.common.Progress.trace
import java.io.File
import java.io.Serializable

class Kb(dirName: String?) : KB("SUMO", dirName), Serializable {

    lateinit var filenames: Array<String>
        private set

    fun make(full: Boolean) : Int {
        return make(getFiles(kbDir, full))
    }

    fun make(files: Array<String>?) : Int {
        filenames = files ?: getFiles(kbDir, true)
        val filePaths: Array<String> = Array(filenames.size) {
            kbDir + File.separatorChar + filenames[it]
        }
        makeKB(this, filePaths)
        return filenames.size
    }

    fun makeClausalForms() {
        var count: Long = 0
        for (fs in formulaIndex.values) {
            for (f in fs) {
                f.getClausalForms()
                if (!silent) {
                    if ((count++ % 100L) == 1L) {
                        println()
                    }
                    print('.')
                }
            }
        }
    }

    companion object {

        private val CORE_FILES = arrayOf("Merge.kif", "Mid-level-ontology.kif", "english_format.kif")

        private val silent = System.getProperties().containsKey("SILENT")

        private fun makeKB(kb: KB, filePaths: Array<String>) {
            filePaths
                .withIndex()
                .forEach { (i, f) ->
                    //System.err.println(basename(f))
                    kb.addConstituent(f)
                    trace(i.toLong())
                }
        }

        private fun getFiles(dirName: String, full: Boolean): Array<String> {
            if (full) {
                val list: MutableList<String> = ArrayList(listOf<String>(*CORE_FILES))
                getKifs(dirName)
                    ?.filterNot { list.contains(it) }
                    ?.forEach {
                        list.add(it)
                    }
                return list.toTypedArray()
            }
            return CORE_FILES
        }

        private fun getKifs(dirName: String): Array<String>? {
            val file = File(dirName)
            if (file.exists() && file.isDirectory()) {
                return file.list { dir: File, name: String -> name.endsWith(".kif") }
            }
            return arrayOf<String>()
        }
    }
}
